### PR TITLE
[R4R] add ics23 proof support

### PR DESCRIPTION
**Pull request by yutianwu**
This pr will add ics23 proof to the multistore query. 

The new query path `ics23-key` will return ics23 proof if proof is requested.

Example ics23 proof query:

```go 
	query := abci.RequestQuery{
		Path: "/ibc/ics23-key", 
		Data: []byte("key"), 
		Height: ver,
		Prove: true,
	}
```

Also, we change the hash calculation of the root multistore (aka app hash), the reason is we can apply the ics23 proof directly by doing this.

## Reference

https://github.com/cosmos/cosmos-sdk/blob/c303d7f3417f67818bc1d2dce6b53732b0db81e4/store/types/commit_info.go#L23

### CODE REVIEW

File: store/proofs/create.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
+package proofs
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+
+	"github.com/tendermint/tendermint/crypto/merkle"
```
Comment (guagualvcha) [inline store/proofs/create.go:8]: create.go seems not a good name.

**Comment by owen-reorg** [comment]
LGTM

**Review by owen-reorg — APPROVED** [review]

File: store/iavlstore.go [hunk 2]
```diff
@@ -261,6 +288,36 @@ func (st *IavlStore) Query(req abci.RequestQuery) (res abci.ResponseQuery) {
...
+		commitmentProof *ics23.CommitmentProof
+		err             error
+	)
+
+	if exists {
+		// value was found
+		commitmentProof, err = tree.GetMembershipProof(key)
+		if err != nil {
+			// sanity check: If value was found, membership proof must be creatable
+			panic(fmt.Sprintf("unexpected value for empty proof: %s", err.Error()))
```
Comment (guagualvcha) [inline store/iavlstore.go:305]: Can we not panic but return an error instead?

File: store/proof.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+		Core: StoreCore{
+			CommitID: CommitID{
+				Version: version,
+				Hash:    iavlRoot[0],
+			},
+		},
+	}
+	storeHash := storeInfo.Hash()
+
+	simplePo := poz[1]
```
Comment (guagualvcha) [inline store/proof.go:186]: Please add `if iavlPo.ProofOp().Type != ProofOpSimpleMerkleCommitment `

File: store/iavlstore.go [hunk 4]
```diff
@@ -244,6 +245,32 @@ func (st *IavlStore) Query(req abci.RequestQuery) (res abci.ResponseQuery) {
...
+		if err != nil {
+			// sanity check: If value for given version was retrieved, immutable tree must also be retrievable
+			panic(fmt.Sprintf("version exists in store but could not retrieve corresponding versioned tree in store, %s", err.Error()))
+		}
+		mtree := &iavl.MutableTree{
+			ImmutableTree: iTree,
+		}
+
+		// get proof from tree and convert to merkle.Proof before adding to result
+		res.Proof = getProofFromTree(mtree, req.Data, res.Value != nil)
```
Comment (owen-reorg) [inline store/iavlstore.go:273]: Where do we return the store commit version for composing proof in BSC side?

Comment (yutianwu) [reply store/iavlstore.go:273]: the version is the height itself, which is passed by the querier

File: store/iavlstore.go (hunk 2 above)
Comment (yutianwu) [reply store/iavlstore.go:305]: that's fine for there are still some panics

File: store/proof.go (hunk 3 above)
Comment (yutianwu) [reply store/proof.go:186]: you are right

File: store/proofs/create.go (hunk 1 above)
Comment (yutianwu) [reply store/proofs/create.go:8]: I just copied the files from tendermint, so I think we can just keep these files for audit or further changes.

File: go.mod [hunk 5]
```diff
@@ -59,19 +60,20 @@ require (
...
+	gopkg.in/yaml.v2 v2.4.0 // indirect
+	gopkg.in/yaml.v3 v3.0.1 // indirect
 )
 
 replace (
-	github.com/tendermint/go-amino => github.com/bnb-chain/bnc-go-amino v0.14.1-binance.2
-	github.com/tendermint/iavl => github.com/bnb-chain/bnc-tendermint-iavl v0.12.0-binance.4
-	github.com/tendermint/tendermint => github.com/bnb-chain/bnc-tendermint v0.32.3-binance.7
-	github.com/zondax/ledger-cosmos-go => github.com/bnb-chain/ledger-cosmos-go v0.9.9-binance.3
+	github.com/tendermint/go-amino => github.com/binance-chain/bnc-go-amino v0.14.1-binance.2
```
Comment (guagualvcha) [inline go.mod:74]: why it have to be `binance-chain`?

File: store/iavlstore.go (hunk 4 above)
Comment (owen-reorg) [reply store/iavlstore.go:273]: I doubt it.
```
type CommitInfo struct {
	Version int64
	StoreInfos []StoreInfo
}

type StoreInfo struct {
	Name string
	Core StoreCore
}

type StoreCore struct {
	CommitID CommitID
}

type CommitID struct {
	Version int64
	Hash    []byte
}
``` 

The version in CommitInfo is the height, but what we need is the one in CommitID.
I didn't check the details yet, but I guess it should be the latest change height of the store.
It's better to double-check that.

**Review by guagualvcha — APPROVED** [review]

**Comment by guagualvcha** [comment]
LGTM

**Review by guagualvcha — APPROVED** [review]

**Pull request merged**